# Capstone Project - The Battle of Neighborhoods


## Introduction/Business Problem
### Introduction where you discuss the business problem and who would be interested in this project.


"Would you recommend a location in Norway to open a new shop". in the city center
This shop can be Coffe shop, Bar, Pizza Restaurant,	Grocery Store, Sushi Restaurant etc.
Let's say we want to open a new shop in Norway.
As company's new business, we can ask two questions.
-What type of shop, we can open (type of shop)
-Where we should open this shop(Location)
We can answer that in customer point of view, most of shops  is a part of whole day or night activities. shop should has many restaurants and shopping places nearby. Transportation is also an important factor. Customer can walk to shop within 5 minutes from public transport facilities such as bus stop and metro station.
We want to concentrate on selection of shop  location according to its nearby environment. shop facility and rental price is not concern. 
I work with my teammates and worling with the best possible locations to build the new shop. Which location should be suggested to the stakeholder ?

### data

Data where you describe the data that will be used to solve the problem and the source of the data.

According to the question, I need to find following data to resolve the problem.

1. List of shopping centres in Norway in all cities.
We can find a list of  shopping centres in Norway, from wikipedia from this link
'https://en.wikipedia.org/wiki/List_of_shopping_centres_in_Norway'

2. Geographic coordinate of those shopping centres in Norway.
We can find a list of Geographic coordinate of those shopping centre, from this link
'https://www.geodatos.net/en/coordinates/norway'


Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [7]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.18.1-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00  21.56 MB/s
geopy-1.18.1-p 100% |################################| Time: 0:00:00  32.86 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  52.63 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  28.30 MB/s
vincent-0.4.4- 100% |###################

##  Download and Explore Dataset

Let's now explore the shopping centres in Norway

In [8]:
import numpy as np  
import pandas as pd 
from bs4 import BeautifulSoup

import requests

url='https://en.wikipedia.org/wiki/List_of_shopping_centres_in_Norway'
df = pd.read_html(url, header=0, flavor = 'bs4')
df1=df[0]
df1

Name                 Location  Number of stores
0       Strømmen Storsenter                 Strømmen               201
1          Sørlandssenteret             Kristiansand               200
2       Sandvika Storsenter                 Sandvika               195
3                  Amfi Moa                  Ålesund               185
4      Oslo City / Byporten      Oslo, Bydel Sentrum               175
5                   Kvadrat                  Sandnes               160
6            Ski Storsenter                      Ski               145
7       Jessheim Storsenter                 Jessheim               144
8                     Arken                   Bergen               138
9        Lagunen Storsenter                   Bergen               135
10        Sartor Storsenter                    Fjell               134
11           Farmandstredet                 Tønsberg               126
12         Gulskogen Senter                  Drammen               125
13         Storo Storsenter       Oslo, Bydel Sagene               125
14         Jekta Storsenter                   Tromsø               130
15             Metro Senter                Lørenskog               102
16                AMFI Moss                     Moss               100
17           AMFI Steinkjer                Steinkjer               100
18               Liertoppen                  Drammen               100
19           Stovner Senter      Oslo, Bydel Stovner               100
20      Buskerud Storsenter             Krokstadelva                90
21               Vestkanten                   Bergen                90
22                 Torvbyen              Fredrikstad                85
23                CC Gjøvik                   Gjøvik                83
24     Lørenskog Storsenter                Lørenskog                83
25                 Byporten      Oslo, Bydel Sentrum                82
26                Down Town                Porsgrunn                81
27          Lillestrøm Torv               Lillestrøm                80
28         Oasen Storsenter                   Karmøy                80
29                Trekanten                    Asker                80
30                  CC Vest       Oslo, Bydel Ullern                78
31                 Herkules                    Skien                78
32                AMFI Borg                Sarpsborg                75
33                 Storbyen                Sarpsborg                75
34         Vinterbro Senter                Vinterbro                75
35                City Nord                     Bodø                74
36           Trondheim Torg                Trondheim                73
37            Tveita Senter         Oslo, Bydel Alna                73
38        Bergen Storsenter                   Bergen                72
39                 City Syd                Trondheim                71
40              Aker Brygge      Oslo, Bydel Frogner                70
41                Galleriet                   Bergen                70
42     Stavanger Storsenter                Stavanger                70
43            Østfoldhallen              Fredrikstad                70
44    Maxi Hamar Storsenter                    Hamar                68
45        Amanda Storsenter                Haugesund                67
46                Magasinet                  Drammen                65
47             Strandtorget              Lillehammer                65
48                   Roseby                    Molde                64
49         Manglerud Senter     Oslo, Bydel Østensjø                63
50               AMFI Vågen                Stavanger                60
51                   Kilden                Stavanger                60
52           Knarvik Senter                   Bergen                60
53                 Solsiden                Trondheim                60
54          Linderud Senter       Oslo, Bydel Bjerke                58
55              Mosseporten                    

In [9]:
print('the number of the most famous city center in Norway are {}'.format(len(df1)))

the number of the most famous city center in Norway are 69


We can group location 

In [10]:
df1 = df1.groupby(['Location'])['Name'].apply(', '.join).reset_index()

df1

Location                                               Name
0                     Asker                           Trekanten, Holmensentret
1                    Bergen  Arken, Lagunen Storsenter, Vestkanten, Bergen ...
2                      Bodø                                          City Nord
3                   Drammen            Gulskogen Senter, Liertoppen, Magasinet
4                     Fjell                                  Sartor Storsenter
5               Fredrikstad                            Torvbyen, Østfoldhallen
6                    Gjøvik                                          CC Gjøvik
7                     Hamar                  Maxi Hamar Storsenter, Kvartal 48
8                 Haugesund                                  Amanda Storsenter
9                  Jessheim                                Jessheim Storsenter
10                   Karmøy                                   Oasen Storsenter
11             Kristiansand                                   Sørlandssenteret
12             Krokstadelva                                Buskerud Storsenter
13              Lillehammer                                       Strandtorget
14               Lillestrøm                                    Lillestrøm Torv
15                Lørenskog                 Metro Senter, Lørenskog Storsenter
16                    Molde                                             Roseby
17                     Moss                             AMFI Moss, Mosseporten
18                   Namsos                                  Namsos Storsenter
19                   Narvik                                        AMFI Narvik
20                 Notodden                                      Tuvensenteret
21                     Oslo                                        Norway Shop
22         Oslo, Bydel Alna                                      Tveita Senter
23       Oslo, Bydel Bjerke                                    Linderud Senter
24      Oslo, Bydel Frogner                                        Aker Brygge
25  Oslo, Bydel Grünerløkka                            Gunerius Shoppingsenter
26       Oslo, Bydel Sagene                                   Storo Storsenter
27      Oslo, Bydel Sentrum                     Oslo City / Byporten, Byporten
28      Oslo, Bydel Stovner                                     Stovner Senter
29       Oslo, Bydel Ullern                                            CC Vest
30     Oslo, Bydel Østensjø                                   Manglerud Senter
31                Porsgrunn                                          Down Town
32                  Sandnes                                            Kvadrat
33                 Sandvika                                Sandvika Storsenter
34                Sarpsborg                                AMFI Borg, Storbyen
35                      Ski                                     Ski Storsenter
36                    Skien                                           Herkules
37                Stavanger           Stavanger Storsenter, AMFI Vågen, Kilden
38                Steinkjer                                     AMFI Steinkjer
39                 Strømmen                                Strømmen Storsenter
40                   Tromsø  Jekta Storsenter, Nerstranda, AMFI Pyramiden, ...
41                Trondheim  Trondheim Torg, City Syd, Solsiden, KBS Kjøpes...
42                 Tønsberg                                     Farmandstredet
43                Vinterbro                                   Vinterbro Senter
44                  Ålesund                       Amfi Moa, Ålesund Storsenter

In [11]:
print('the number of the most famous city center group by Location in Norway are {}'.format(len(df1)))

the number of the most famous city center group by Location in Norway are 45


Let's now explore Geographic coordinate of those shopping centres in norway

In [12]:
import pandas as pd
url2='https://www.geodatos.net/en/coordinates/norway'
df0 = pd.read_html(url2, header=0, flavor = 'bs4')
df0[1].head()
df2=df0[1]
df2.reset_index(inplace=True)
df2

index          City             Coordinates
0       0          Oslo  59.9127312, 10.7460899
1       1        Bergen   60.3929901, 5.3241501
2       2     Trondheim  63.4304886, 10.3950596
3       3     Stavanger   58.9700508, 5.7333202
4       4       Drammen  59.7438889, 10.2044897
5       5   Fredrikstad     59.2181015, 10.9298
6       6  Kristiansand   58.1467094, 7.9956002
7       7       Sandnes   58.8524399, 5.7352099
8       8        Tromsø    69.6489029, 18.95508
9       9     Sarpsborg  59.2839088, 11.1096201
10     10         Skien   59.2096214, 9.6089697
11     11       Ålesund   62.4722519, 6.1549201
12     12    Sandefjord    59.1311798, 10.21665
13     13     Haugesund   59.4137802, 5.2680001
14     14      Tønsberg    59.26754, 10.4076204
15     15          Moss  59.4340286, 10.6577101
16     16     Porsgrunn   59.1405411, 9.6561003
17     17          Bodø  67.2799988, 14.4050102
18     18       Arendal   58.4615097, 8.7725296
19     19         Hamar  60.7944984, 11.0679798
20     20     Ytrebygda   60.3050385, 5.2823601

We can also get Geographic coordinate of those shopping centres in norway from CSV file

In [13]:
# The code was removed by Watson Studio for sharing.

Location   Latitude  Longitude
0                Oslo  59.912731  10.746090
1              Bergen  60.392990   5.324150
2           Trondheim  63.430489  10.395060
3           Stavanger  58.970051   5.733320
4             Drammen  59.743889  10.204490
5         Fredrikstad  59.218102  10.929800
6        Kristiansand  58.146709   7.995600
7             Sandnes  58.852440   5.735210
8              Tromsø  69.648903  18.955080
9           Sarpsborg  59.283909  11.109620
10              Skien  59.209621   9.608970
11            Ålesund  62.472252   6.154920
12         Sandefjord  59.131180  10.216650
13          Haugesund  59.413780   5.268000
14           Tønsberg  59.267540  10.407620
15               Moss  59.434029  10.657710
16          Porsgrunn  59.140541   9.656100
17               Bodø  67.279999  14.405010
18            Arendal  58.461510   8.772530
19              Hamar  60.794498  11.067980
20          Vinterbro  60.305039   5.282360
21              Asker  59.833328  10.437210
22              Fjell  60.089668  11.472220
23             Gjøvik  60.795430  10.691630
24       Krokstadelva  59.759138  10.010187
25        Lillehammer  61.115139  10.466280
26         Lillestrøm  59.955970  11.050379
27          Lørenskog  59.883227  10.955149
28              Molde  62.737235   7.160731
29             Namsos  64.466240  11.495720
30             Narvik  68.438377  17.427200
31           Notodden  59.559229   9.262836
32           Sandvika  59.890905  10.527787
33                Ski  59.735206  10.908211
34          Steinkjer  64.015019  11.495263
35           Strømmen  59.945765  11.007263
36             Karmøy  59.280102   5.307111
37           Jessheim  60.141510  11.175150
38             Larvik  59.053280  10.035170
39           Notodden  59.559360   9.258530
40  Halden.59.1247787  11.387540        NaN
41             Verdal  63.793320  11.481700
42               Alta  69.968872  23.271650
43            Elverum  60.881908  11.562310
44              Søgne  58.093281   7.782940

Let's merge now both the data frame 

In [14]:

from functools import reduce
#df1 = pd.read_csv(r'cord.csv', names=['Location','Latitude','Longitude'] ,encoding='latin-1')

df2=df2.sort_values(by=['Location'])
df3 = pd.merge(df1, df2, on= "Location", how= "inner")

df3


Location                                               Name  \
0          Asker                           Trekanten, Holmensentret   
1         Bergen  Arken, Lagunen Storsenter, Vestkanten, Bergen ...   
2           Bodø                                          City Nord   
3        Drammen            Gulskogen Senter, Liertoppen, Magasinet   
4          Fjell                                  Sartor Storsenter   
5    Fredrikstad                            Torvbyen, Østfoldhallen   
6         Gjøvik                                          CC Gjøvik   
7          Hamar                  Maxi Hamar Storsenter, Kvartal 48   
8      Haugesund                                  Amanda Storsenter   
9       Jessheim                                Jessheim Storsenter   
10        Karmøy                                   Oasen Storsenter   
11  Kristiansand                                   Sørlandssenteret   
12  Krokstadelva                                Buskerud Storsenter   
13   Lillehammer                                       Strandtorget   
14    Lillestrøm                                    Lillestrøm Torv   
15     Lørenskog                 Metro Senter, Lørenskog Storsenter   
16         Molde                                             Roseby   
17          Moss                             AMFI Moss, Mosseporten   
18        Namsos                                  Namsos Storsenter   
19        Narvik                                        AMFI Narvik   
20      Notodden                                      Tuvensenteret   
21      Notodden                                      Tuvensenteret   
22          Oslo                                        Norway Shop   
23     Porsgrunn                                          Down Town   
24       Sandnes                                            Kvadrat   
25      Sandvika                                Sandvika Storsenter   
26     Sarpsborg                                AMFI Borg, Storbyen   
27           Ski                                     Ski Storsenter   
28         Skien                                           Herkules   
29     Stavanger           Stavanger Storsenter, AMFI Vågen, Kilden   
30     Steinkjer                                     AMFI Steinkjer   
31      Strømmen                                Strømmen Storsenter   
32        Tromsø  Jekta Storsenter, Nerstranda, AMFI Pyramiden, ...   
33     Trondheim  Trondheim Torg, City Syd, Solsiden, KBS Kjøpes...   
34      Tønsberg                                     Farmandstredet   
35     Vinterbro                                   Vinterbro Senter   
36       Ålesund                       Amfi Moa, Ålesund Storsenter   

     Latitude  Longitude  
0   59.833328  10.437210  
1   60.392990   5.324150  
2   67.279999  14.405010  
3   59.743889  10.204490  
4   60.089668  11.472220  
5   59.218102  10.929800  
6   60.795430  10.691630  
7   60.794498  11.067980  
8   59.413780   5.268000  
9   60.141510  11.175150  
10  59.280102   5.307111  
11  58.146709   7.995600  
12  59.759138  10.010187  
13  61.115139  10.466280  
14  59.955970  11.050379  
15  59.883227  10.955149  
16  62.737235   7.160731  
17  59.434029  10.657710  
18  64.466240  11.495720  
19  68.438377  17.427200  
20  59.559229   9.262836  
21  59.559360   9.258530  
22  59.912731  10.746090  
23  59.140541   9.656100  
24  58.852440   5.735210  
25  59.890905  10.527787  
26  59.283909  11.109620  
27  59.735206  10.908211  
28  59.209621   9.608970  
29  58.970051   5.733320  
30  64.015019  11.495263  
31  59.945765  11.007263  
32  69.648903  18.955080  
33  63.430489  10.395060  
34  59.267540  10.407620  
35  60.305039   5.282360  
36  62.472252   6.154920

The recommended shop location needs to have many resturands and shopping venues nearby. Convenient public transport is also required.
These data can be found by using FourSquare API to find these venues around the location. The radius of exploration distance is set to 500 meters, which is about 5 minutes walking distance.

Following type of venue category will be used to search

In [15]:
fs_categories = {
    'Food': '4d4b7105d754a06374d81259',
    'Shop & Service': '4d4b7105d754a06378d81259',
    'Bus Stop': '52f2ab2ebcbc57f1066b8b4f',
    'Metro Station': '4bf58dd8d48988d1fd931735',
    'Nightlife Spot': '4d4b7105d754a06376d81259',
    'Arts & Entertainment': '4d4b7104d754a06370d81259'
}

In [16]:
', '.join([ cat for cat in fs_categories])


'Bus Stop, Food, Nightlife Spot, Shop & Service, Metro Station, Arts & Entertainment'

### Install FourSquare client library

In [17]:

!pip install foursquare


Requirement not upgraded as not directly required: requests>=2.1 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from foursquare)
Requirement not upgraded as not directly required: six in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from foursquare)
Requirement not upgraded as not directly required: chardet<3.1.0,>=3.0.2 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests>=2.1->foursquare)
Requirement not upgraded as not directly required: idna<2.7,>=2.5 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests>=2.1->foursquare)
Requirement not upgraded as not directly required: urllib3<1.23,>=1.21.1 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests>=2.1->foursquare)
Requirement not upgraded as not directly required: certifi>=2017.4.17 in /opt/conda/envs/DSX-Python35/lib/python3.5/site-packages (from requests>=2.1->foursquare)
  Running setup.py bdist_wheel for foursquare ... done
  Stored in dir

Using Foursquare API, using my credentails

In [18]:
CLIENT_ID = 'NZX34ZUR3FS3UV5SF5TD2EI4RIUBO43EKXZB2CEHA4KW1UZL' # your Foursquare ID
CLIENT_SECRET = 'X0N1JGX2W4MUU1GCA0ZYHXMSKIBZZ3GRHYKBD3K0D1VKX5TL' # your Foursquare Secret
VERSION = '20181230' # Foursquare API version

print('My credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

My credentails:
CLIENT_ID: NZX34ZUR3FS3UV5SF5TD2EI4RIUBO43EKXZB2CEHA4KW1UZL
CLIENT_SECRET:X0N1JGX2W4MUU1GCA0ZYHXMSKIBZZ3GRHYKBD3K0D1VKX5TL


Use the first shopping center (Asker) in the list as example to explore venues nearyby

In [19]:
city_latitude = df3.loc[0, 'Latitude'] # neighborhood latitude value
city_longitude = df3.loc[0, 'Longitude'] # neighborhood longitude value

city_name = df3.loc[0, 'Location'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(city_name, 
                                                               city_latitude, 
                                                               city_longitude))

Latitude and longitude values of Asker are 59.833328200000004, 10.4372101.


Let's  create map for Asker sopping center


In [20]:
# create map of New York using latitude and longitude values
Latitude=59.8333282
Longitude=10.437210
Location='Asker'
map_norway = folium.Map(location=[Latitude, Longitude], zoom_start=10)

# add markers to map
for lat, lng, city in zip(df3['Latitude'], df3['Longitude'],  df3['Location']):
    label = ' {}'.format(Location)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_norway)  
    
map_norway

In [21]:
LIMIT = 100 # limit of number of venues returned by Foursquare API



radius = 500 # define radius



# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    city_latitude, 
    city_longitude, 
    radius, 
    LIMIT)
url 

'https://api.foursquare.com/v2/venues/explore?&client_id=NZX34ZUR3FS3UV5SF5TD2EI4RIUBO43EKXZB2CEHA4KW1UZL&client_secret=X0N1JGX2W4MUU1GCA0ZYHXMSKIBZZ3GRHYKBD3K0D1VKX5TL&v=20181230&ll=59.833328200000004,10.4372101&radius=500&limit=100'

Send the GET request and examine the resutls

In [22]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5c2a79476a607133f4cd451c'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4c2f8b7f7cc0c9b673ebeb9a-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/arts_entertainment/movietheater_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d17f941735',
         'name': 'Movie Theater',
         'pluralName': 'Movie Theaters',
         'primary': True,
         'shortName': 'Movie Theater'}],
       'id': '4c2f8b7f7cc0c9b673ebeb9a',
       'location': {'address': 'Kinoveien 59',
        'cc': 'NO',
        'city': 'Asker',
        'country': 'Norge',
        'distance': 369,
        'formattedAddress': ['Kinoveien 59', '1338 Asker', 'Norge'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 59.835846193478005,
      

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [23]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [24]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name     categories        lat        lng
0       Asker Kino  Movie Theater  59.835846  10.432897
1  Asker Kulturhus   Concert Hall  59.835837  10.432828
2      Ultra Asker  Grocery Store  59.832332  10.429760
3  Kårner Kaffebar    Coffee Shop  59.834795  10.434470
4  Kaffebrenneriet    Coffee Shop  59.835696  10.432850

And how many venues were returned by Foursquare?

In [25]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

16 venues were returned by Foursquare.


#### Let's create a function to repeat the same process to all the shopping centers in Norway

In [26]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['city_name', 
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *city_venues*.

In [27]:
# type your answer here
norway_venues = getNearbyVenues(names=df3['Location'],
                                   latitudes=df3['Latitude'],
                                   longitudes=df3['Longitude']
                                  )

Asker
Bergen
Bodø
Drammen
Fjell
Fredrikstad
Gjøvik
Hamar
Haugesund
Jessheim
Karmøy
Kristiansand
Krokstadelva
Lillehammer
Lillestrøm
Lørenskog
Molde
Moss
Namsos
Narvik
Notodden
Notodden
Oslo
Porsgrunn
Sandnes
Sandvika
Sarpsborg
Ski
Skien
Stavanger
Steinkjer
Strømmen
Tromsø
Trondheim
Tønsberg
Vinterbro
Ålesund


#### Let's check the size of the resulting dataframe

In [28]:
print(norway_venues.shape)
norway_venues.head()

(843, 7)


city_name   Latitude  Longitude            Venue  Venue Latitude  \
0     Asker  59.833328   10.43721       Asker Kino       59.835846   
1     Asker  59.833328   10.43721  Asker Kulturhus       59.835837   
2     Asker  59.833328   10.43721      Ultra Asker       59.832332   
3     Asker  59.833328   10.43721  Kårner Kaffebar       59.834795   
4     Asker  59.833328   10.43721  Kaffebrenneriet       59.835696   

   Venue Longitude Venue Category  
0        10.432897  Movie Theater  
1        10.432828   Concert Hall  
2        10.429760  Grocery Store  
3        10.434470    Coffee Shop  
4        10.432850    Coffee Shop

Let's check how many venues were returned for each city centre

In [29]:
norway_venues.groupby('city_name').count()

Latitude  Longitude  Venue  Venue Latitude  Venue Longitude  \
city_name                                                                   
Asker               16         16     16              16               16   
Bergen             100        100    100             100              100   
Bodø                 5          5      5               5                5   
Drammen             26         26     26              26               26   
Fredrikstad          5          5      5               5                5   
Gjøvik              22         22     22              22               22   
Hamar               20         20     20              20               20   
Haugesund           23         23     23              23               23   
Jessheim             9          9      9               9                9   
Karmøy               7          7      7               7                7   
Kristiansand        48         48     48              48               48   
Krokstadelva         4          4      4               4                4   
Lillehammer         19         19     19              19               19   
Lillestrøm          25         25     25              25               25   
Molde               14         14     14              14               14   
Moss                 9          9      9               9                9   
Namsos               6          6      6               6                6   
Narvik              12         12     12              12               12   
Notodden             9          9      9               9                9   
Oslo               100        100    100             100              100   
Porsgrunn            4          4      4               4                4   
Sandnes             28         28     28              28               28   
Sandvika            23         23     23              23               23   
Sarpsborg            7          7      7               7                7   
Skien               10         10     10              10               10   
Stavanger           74         74     74              74               74   
Steinkjer            8          8      8               8                8   
Strømmen            16         16     16              16               16   
Tromsø              80         80     80              80               80   
Trondheim           47         47     47              47               47   
Tønsberg            27         27     27              27               27   
Vinterbro            4          4      4               4                4   
Ålesund             36         36     36              36               36   

              Venue Category  
city_name                     
Asker                     16  
Bergen                   100  
Bodø                       5  
Drammen                   26  
Fredrikstad                5  
Gjøvik                    22  
Hamar                     20  
Haugesund                 23  
Jessheim                   9  
Karmøy                     7  
Kristiansand              48  
Krokstadelva               4  
Lillehammer               19  
Lillestrøm                25  
Molde                     14  
Moss                       9  
Namsos                     6  
Narvik                    12  
Notodden                   9  
Oslo                     100  
Porsgrunn                  4  
Sandnes                   28  
Sandvika                  23  
Sarpsborg                  7  
Skien                     10  
Stavanger                 74  
Steinkjer                  8  
Strømmen                  16  
Tromsø                    80  
Trondheim                 47  
Tønsberg                  27  
Vinterbro                  4  
Ålesund                   36

#### Let's find out how many unique categories can be curated from all the returned venues

In [30]:
print('There are {} uniques categories.'.format(len(norway_venues['Venue Category'].unique())))

There are 141 uniques categories.


## 3. Analyze Each City center

In [31]:
# one hot encoding
norway_onehot = pd.get_dummies(norway_venues[['Venue Category']], prefix="", prefix_sep="")

# add  city_ name column back to dataframe
norway_onehot['city_name'] = norway_venues['city_name'] 

# move neighborhood column to the first column
fixed_columns = [norway_onehot.columns[-1]] + list(norway_onehot.columns[:-1])
norway_onehot = norway_onehot[fixed_columns]

norway_onehot.head()

city_name  American Restaurant  Art Gallery  Art Museum  \
0     Asker                    0            0           0   
1     Asker                    0            0           0   
2     Asker                    0            0           0   
3     Asker                    0            0           0   
4     Asker                    0            0           0   

   Arts & Crafts Store  Asian Restaurant  BBQ Joint  Bakery  Bar  Beach  \
0                    0                 0          0       0    0      0   
1                    0                 0          0       0    0      0   
2                    0                 0          0       0    0      0   
3                    0                 0          0       0    0      0   
4                    0                 0          0       0    0      0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Bistro  Boat or Ferry  \
0                0         0            0           0       0              0   
1                0         0            0           0       0              0   
2                0         0            0           0       0              0   
3                0         0            0           0       0              0   
4                0         0            0           0       0              0   

   Bookstore  Brazilian Restaurant  Breakfast Spot  Brewery  Bridge  \
0          0                     0               0        0       0   
1          0                     0               0        0       0   
2          0                     0               0        0       0   
3          0                     0               0        0       0   
4          0                     0               0        0       0   

   Burger Joint  Burrito Place  Bus Station  Café  Camera Store  \
0             0              0            0     0             0   
1             0              0            0     0             0   
2             0              0            0     0             0   
3             0              0            0     0             0   
4             0              0            0     0             0   

   Caribbean Restaurant  Chinese Restaurant  Church  Clothing Store  \
0                     0                   0       0               0   
1                     0                   0       0               0   
2                     0                   0       0               0   
3                     0                   0       0               0   
4                     0                   0       0               0   

   Cocktail Bar  Coffee Shop  Comedy Club  Comfort Food Restaurant  \
0             0            0            0                        0   
1             0            0            0                        0   
2             0            0            0                        0   
3             0            1            0                        0   
4             0            1            0                        0   

   Concert Hall  Construction & Landscaping  Convenience Store  \
0             0                           0                  0   
1             1                           0                  0   
2             0                           0                  0   
3             0                           0                  0   
4             0                           0                  0   

   Coworking Space  Creperie  Deli / Bodega  Department Store  Dessert Shop  \
0                0         0              0                 0             0   
1                0         0              0                 0             0   
2                0         0              0                 0             0   
3                0         0              0                 0             0   
4                0         0              0                 0             0   

   Diner  Dive Bar  Electronics Store  Event Space  Falafel Restaurant  \
0      0         0                  0            0                   0   
1    

And let's examine the new dataframe size.

In [32]:
norway_onehot.shape

(843, 142)

#### Next, let's group rows by city name and by taking the mean of the frequency of occurrence of each category

In [33]:
norway_grouped = norway_onehot.groupby('city_name').mean().reset_index()
norway_grouped

city_name  American Restaurant  Art Gallery  Art Museum  \
0          Asker             0.000000         0.00    0.000000   
1         Bergen             0.030000         0.03    0.000000   
2           Bodø             0.000000         0.00    0.000000   
3        Drammen             0.038462         0.00    0.000000   
4    Fredrikstad             0.000000         0.00    0.000000   
5         Gjøvik             0.000000         0.00    0.000000   
6          Hamar             0.050000         0.00    0.000000   
7      Haugesund             0.000000         0.00    0.000000   
8       Jessheim             0.111111         0.00    0.000000   
9         Karmøy             0.000000         0.00    0.000000   
10  Kristiansand             0.000000         0.00    0.020833   
11  Krokstadelva             0.000000         0.00    0.000000   
12   Lillehammer             0.000000         0.00    0.000000   
13    Lillestrøm             0.000000         0.00    0.000000   
14         Molde             0.000000         0.00    0.000000   
15          Moss             0.000000         0.00    0.000000   
16        Namsos             0.000000         0.00    0.000000   
17        Narvik             0.000000         0.00    0.000000   
18      Notodden             0.000000         0.00    0.000000   
19          Oslo             0.000000         0.00    0.000000   
20     Porsgrunn             0.000000         0.00    0.000000   
21       Sandnes             0.000000         0.00    0.000000   
22      Sandvika             0.000000         0.00    0.000000   
23     Sarpsborg             0.000000         0.00    0.000000   
24         Skien             0.000000         0.00    0.000000   
25     Stavanger             0.000000         0.00    0.000000   
26     Steinkjer             0.000000         0.00    0.000000   
27      Strømmen             0.000000         0.00    0.000000   
28        Tromsø             0.037500         0.00    0.037500   
29     Trondheim             0.000000         0.00    0.000000   
30      Tønsberg             0.037037         0.00    0.000000   
31     Vinterbro             0.000000         0.00    0.000000   
32       Ålesund             0.027778         0.00    0.027778   

    Arts & Crafts Store  Asian Restaurant  BBQ Joint    Bakery       Bar  \
0                0.0000          0.000000   0.000000  0.000000  0.000000   
1                0.0000          0.010000   0.000000  0.040000  0.120000   
2                0.0000          0.000000   0.000000  0.000000  0.000000   
3                0.0000          0.000000   0.000000  0.038462  0.038462   
4                0.0000          0.000000   0.000000  0.000000  0.000000   
5                0.0000          0.000000   0.000000  0.000000  0.000000   
6                0.0000          0.000000   0.050000  0.000000  0.050000   
7                0.0000          0.043478   0.000000  0.043478  0.086957   
8                0.0000          0.000000   0.000000  0.000000  0.000000   
9                0.0000          0.000000   0.000000  0.000000  0.000000   
10               0.0000          0.000000   0.000000  0.000000  0.020833   
11               0.0000          0.250000   0.000000  0.000000  0.000000   
12               0.0000          0.000000   0.000000  0.000000  0.052632   
13               0.0000          0.000000   0.000000  0.000000  0.040000   
14               0.0000          0.000000   0.000000  0.000000  0.000000   
15               0.0000          0.000000   0.000000  0.111111  0.000000   
16               0.0000          0.000000   0.000000  0.000000  0.000000   
17               0.0000          0.000000   0.000000  0.000000  0.166667   
18               0.0000          0.000000   0.000000  0.000000  0.000000   
19               0.0000          0.010000   0.000000  0.000000  0.060000   
20               0.0000          0.000000   0.000000  0.000000  0.000000   
21               0.0000          0.000000   0.000000  0.035714  0.000000   
22            

#### Let's confirm the new size

In [34]:
norway_grouped.shape

(33, 144)

#### Let's print each city along with the top 5 most common venues

In [35]:
num_top_venues = 5

for hood in norway_grouped['city_name']:
    print("----"+hood+"----")
    temp = norway_grouped[norway_grouped['city_name'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Asker----
              venue  freq
0       Coffee Shop  0.12
1       Pizza Place  0.12
2     Grocery Store  0.12
3     Movie Theater  0.06
4  Sushi Restaurant  0.06


----Bergen----
         venue  freq
0          Bar  0.12
1  Coffee Shop  0.07
2          Pub  0.07
3   Restaurant  0.07
4         Café  0.06


----Bodø----
                 venue  freq
0          Event Space  0.25
1    Convenience Store  0.25
2                 Café  0.25
3        Grocery Store  0.25
4  American Restaurant  0.00


----Drammen----
                  venue  freq
0                 Hotel  0.14
1                 Plaza  0.09
2           Pizza Place  0.09
3           Coffee Shop  0.09
4  Fast Food Restaurant  0.09


----Fredrikstad----
                   venue  freq
0   Gym / Fitness Center   0.2
1    Sporting Goods Shop   0.2
2           Soccer Field   0.2
3  Performing Arts Venue   0.2
4          Grocery Store   0.2


----Gjøvik----
                  venue  freq
0  Fast Food Restaurant  0.16
1            Re

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [34]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each city centre.

In [ ]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['city_name']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common shop'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common shop'.format(ind+1))

# create a new dataframe
norway_venues_sorted = pd.DataFrame(columns=columns)
norway_venues_sorted['city_name'] = norway_grouped['city_name']

for ind in np.arange(norway_grouped.shape[0]):
    norway_venues_sorted.iloc[ind, 1:] = return_most_common_venues(norway_grouped.iloc[ind, :], num_top_venues)

norway_venues_sorted

##  Methodology Used



In order to proceed our research, we need:

The Wikipedia page  url='https://en.wikipedia.org/wiki/List_of_shopping_centres_in_Norway'
was scraped using the BeautifulSoup library to build a pandas dataframe listing the shpping center Norway. The dataframe was cleaned and processed appropriately.

The Wikipedia page  url='ur2='https://www.geodatos.net/en/coordinates/norway'

was scraped using the BeautifulSoup library to build a pandas dataframe listing the geographics coordinates shpping center Norway. The dataframe was cleaned and processed appropriately.

The Foursquare API was then used to query each of the cities for the diifferent shops. The resulting locations for each venue were tabulated and stored in a new dataframe and visualised using Folium.
A better visualisation was performed by obtaining the cluster centers for each group of laundromat and hotel. The cluster centers were plotted on a Folium map, with the circle marker weighted corresponding to the number of venues in the specific cluster.
Results
Our research gives this results.


regarding the previous results we can find many suggestions for opening shop in each city in norway

like for example 


----Asker----


              venue  freq
0       Coffee Shop  0.12
1       Pizza Place  0.12
2     Grocery Store  0.12
3     Movie Theater  0.06
4  Sushi Restaurant  0.06


----Bergen----


         venue  freq
0          Bar  0.12
1  Coffee Shop  0.07
2          Pub  0.07
3   Restaurant  0.07
4         Café  0.06


Discusions
In highly populated areas, & were high paying jobs like banking, entertainment is very popular  to predict the best type of shops to open it  like cafe or resturants. The main reason is because a large concentration of shops  and  alike can be found surrounding city centers. This can be seen in the large clusters shown in the Folium map.
Conclusions
You can invest in either income-producing properties or non-income-producing properties. Any leased property is income producing, and vacant properties are non-income producing. You can still earn a capital return on a non-income producing property, just as you would on an investment in a home.
Buying real estate requires substantial due diligence to ensure that you're getting what you expect after you close.